core/src/main/scala/kafka/controller/ControllerContext.scala

```diff
@@ -431,3 +425,22 @@ class ControllerContext {
     targetState.validPreviousStates.contains(partitionStates(partition))
 
 }
+
+/**
+  * The ControllerContextSnapshot is an immutable snapshot of the ControllorContext.
+  * The motivation for this class is that we don't need to calculate certain fields
```

Comment:
nit: I think we can specifically mention this is meant for the case we don't need recalculation

> The motivation is for this class is that *there are cases* we don't need to calculate certain fields repeatedly, including ..., *and this class is only meant to be used for the cases of avoiding redundant evaluation of the fields*.

Reply:
I feel the current comment is already saying that about avoiding recalculation.